File: SkylakeNAT/encryptor.cpp
#include "env.h"
#include "rc4.h"
#include "md5.h"
#include "encryptor.h"
#include "nedmalloc/memory.h"

#include <string>
#include <sstream>
#include <iostream>

Encryptor::Encryptor(const std::string method, const std::string password)
	: _cipher(NULL)
	, _method(method)
	, _password(password)
{
	initKey(method, password);
}

Encryptor::~Encryptor()
{
	Dispose();
}

void Encryptor::Initialize()
{
	/* initialize OpenSSL */
	OpenSSL_add_all_ciphers();
	OpenSSL_add_all_digests();
	OpenSSL_add_all_algorithms();
	ERR_load_EVP_strings();
	ERR_load_crypto_strings();
}

void Encryptor::Dispose()
{
	_cipher = NULL;
	_context = NULL;
	_key = NULL;
	_encryptCTX = NULL;
	_decryptCTX = NULL;
}

std::shared_ptr<unsigned char> Encryptor::Encrypt(unsigned char* data, int datalen, int& outlen)
{
	outlen = 0;
	if (datalen < 0 || (!data && datalen)) {
		outlen = ~0;
		return NULL;
	}

	if (datalen == 0) {
		return NULL;
	}

	MonitorScope scope(_syncobj);

	// INIT-CTX
	initCipher(_encryptCTX, _iv.get(), _cipher->iv_len, true);

	// ENCR-DATA
	int outLen = datalen + _cipher->block_size;
	std::shared_ptr<unsigned char> cipherText = New(outLen);
	if (EVP_CipherUpdate(_encryptCTX.get(), cipherText.get(), &outLen, data, datalen) <= 0)
	{
		outlen = ~0;
		return NULL;
	}

	outlen = outLen;
	if (cipherText.get()) {
		cipherText.get()[outLen] = '\x0';
	}
	return cipherText;
}

std::shared_ptr<unsigned char> Encryptor::Decrypt(unsigned char* data, int datalen, int& outlen)
{
	outlen = 0;
	if (datalen < 0 || (!data && datalen)) {
		outlen = ~0;
		return NULL;
	}

	if (datalen == 0) {
		return NULL;
	}

	MonitorScope scope(_syncobj);

	// INIT-CTX
	initCipher(_decryptCTX, _iv.get(), _cipher->iv_len, false);

	// DECR-DATA
	int outLen = datalen + _cipher->block_size;
	std::shared_ptr<unsigned char> cipherText = New(outLen);
	if (EVP_CipherUpdate(_decryptCTX.get(), cipherText.get(), &outLen, data, datalen) <= 0) {
		outlen = ~0;
		return NULL;
	}

	outlen = outLen;
	if (cipherText.get()) {
		cipherText.get()[outLen] = '\x0';
	}
	return cipherText;
}

std::shared_ptr<unsigned char> Encryptor::New(int length, int memset)
{
	if (length <= 0)
		return NULL;
	unsigned char* bytes = (unsigned char*)Memory::Alloc((uint32_t)length);
	if (memset >= 0)
		::memset(bytes, 0, _cipher->iv_len);
	return std::shared_ptr<unsigned char>(bytes, [](unsigned char* p) {
		if (p)
			Memory::Free(p);
	});
}

void Encryptor::initCipher(std::shared_ptr<EVP_CIPHER_CTX>& context, const unsigned char* iv, int ivlen, bool isCipher)
{
	bool exception = false;
	do {
		int enc = isCipher ? 1 : 0;
		if (NULL == context.get()) {
			context = std::shared_ptr<EVP_CIPHER_CTX>(
				(EVP_CIPHER_CTX*)OPENSSL_malloc(sizeof(EVP_CIPHER_CTX)), [](EVP_CIPHER_CTX* context) {
				if (context) {
					EVP_CIPHER_CTX_cleanup(context);
					OPENSSL_free(context);
				}
				context = NULL;
			});
			EVP_CIPHER_CTX_init(context.get());
			if (exception = EVP_CipherInit_ex(context.get(), _cipher, NULL, NULL, NULL, enc) <= 0)
				break;
			if (exception = EVP_CIPHER_CTX_set_key_length(context.get(), _cipher->key_len) <= 0)
				break;
			if (exception = EVP_CIPHER_CTX_set_padding(context.get(), 1) <= 0)
				break;
		}
		if (exception = EVP_CipherInit_ex(context.get(), _cipher, NULL, _key.get(), iv, enc) <= 0)
			break;
	} while (0, 0);
	if (exception) {
		context = NULL;
		throw std::runtime_error("There was a problem initializing the cipher that caused an exception to be thrown");
	}
}

void Encryptor::initKey(const std::string& method, const std::string password)
{
	const char* cipherName = method.empty() ? ENCRYPTOR_AES_128_CFB : method.data();
	_cipher = EVP_get_cipherbyname(cipherName);
	if (!_cipher)
		throw std::runtime_error("Such encryption cipher methods are not supported");
	std::shared_ptr<unsigned char> iv = New(_cipher->iv_len, 0);

	_key = New(_cipher->key_len, 0);
	if (EVP_BytesToKey(_cipher, EVP_md5(), NULL, (unsigned char*)password.data(), (int)password.length(),
		1, _key.get(), iv.get()) <= 0) {
		iv = NULL;
		throw std::runtime_error("Bytes to key calculations cannot be performed using cipher with md5(md) key password iv key etc");
	}

	// INIT-IVV
	int ivLen = _cipher->iv_len;
	int md5len = ivLen < sizeof(MD5::HEX) ? sizeof(MD5::HEX) : ivLen;
	_iv = New(ivLen, 0); // RAND_bytes(iv.get(), ivLen);

	std::stringstream ss; // MD5->RC4
	ss << "SkylakeNAT@";
	ss << method;
	ss << ".";
	ss << std::string((char*)_key.get(), _cipher->key_len);
	ss << ".";
	ss << password;
	ComputeMD5(ss.str(), _iv.get(), md5len);
	rc4_crypt(_key.get(), _cipher->key_len, _iv.get(), ivLen, 0, 0);
}

File: SupersocksR/SkylakeNAT/Encryptor.cs
namespace SupersocksR.SkylakeNAT
{
    using System;
    using System.Runtime.InteropServices;
    using System.Security.Cryptography;
    using System.Text;
    using OpenSSL.Core;
    using OpenSSL.Crypto;
    using SupersocksR.Core;

    public unsafe class Encryptor : IDisposable
    {
        public readonly static string[] EncryptionNames = new[]
        {
            "aes-256-cfb",
            "aes-192-cfb",
            "aes-128-cfb",
            "bf-cfb"
        };

        public string method = EncryptionNames[0];
        public string password = string.Empty;
        public byte[] key = BufferSegment.Empty;
        private Cipher cipher = null;
        private IntPtr encryptCTX = IntPtr.Zero;
        private IntPtr decryptCTX = IntPtr.Zero;
        private byte[] iv = null;
        private readonly object LockObj = new object();
        private readonly static byte[] EmptyBuf = new byte[0];

        public void Dispose()
        {
            if (encryptCTX != IntPtr.Zero)
            {
                Native.EVP_CIPHER_CTX_cleanup(encryptCTX);
                Native.OPENSSL_free(encryptCTX);
                encryptCTX = IntPtr.Zero;
            }
            if (decryptCTX != IntPtr.Zero)
            {
                Native.EVP_CIPHER_CTX_cleanup(decryptCTX);
                Native.OPENSSL_free(decryptCTX);
                decryptCTX = IntPtr.Zero;
            }
        }

        ~Encryptor()
        {
            Dispose();
        }

        public Encryptor(string method, string password)
        {
            this.encryptCTX = IntPtr.Zero;
            this.decryptCTX = IntPtr.Zero;
            this.method = method;
            this.password = password;
            this.initKey(method, password);
        }

        private void initCipher(ref IntPtr ctx, byte[] iv, bool isCipher)
        {
            int enc = isCipher ? 1 : 0;
            if (ctx == IntPtr.Zero)
            {
                ctx = Native.OPENSSL_malloc(Marshal.SizeOf(typeof(CipherContext.EVP_CIPHER_CTX)));
                Native.EVP_CIPHER_CTX_init(ctx);
                Native.ExpectSuccess(Native.EVP_CipherInit_ex(ctx, this.cipher.Handle, IntPtr.Zero, null, null, enc));
                Native.ExpectSuccess(Native.EVP_CIPHER_CTX_set_key_length(ctx, key.Length));
                Native.ExpectSuccess(Native.EVP_CIPHER_CTX_set_padding(ctx, 1));
            }
            Native.ExpectSuccess(Native.EVP_CipherInit_ex(ctx, this.cipher.Handle, IntPtr.Zero, key, iv, enc));
        }

        private void initKey(string method, string password)
        {
            cipher = Cipher.CreateByName(method);
            if (cipher == null)
                throw new ArgumentOutOfRangeException("Such encryption cipher methods are not supported");

            byte[] passbuf = Encoding.Default.GetBytes(password);
            key = new byte[cipher.KeyLength];

            iv = new byte[cipher.IVLength];
            if (Native.EVP_BytesToKey(cipher.Handle, MessageDigest.MD5.Handle, null, passbuf, passbuf.Length, 1, key, iv) <= 0)
                throw new ExternalException("Bytes to key calculations cannot be performed using cipher with md5(md) key password iv key etc");

            int ivLen = cipher.IVLength;
            int md5len = ivLen < sizeof(Guid) ? sizeof(Guid) : ivLen;
            iv = new byte[ivLen]; // RAND_bytes(iv.get(), ivLen); = new byte[ivLen]; // RAND_bytes(iv.get(), ivLen);

            // MD5->RC4
            Buffer.BlockCopy(HashAlgorithm<MD5>.
                ComputeHash(
                    merges(
                        Encoding.Default.GetBytes($"SkylakeNAT@{method}."),
                        key,
                        Encoding.Default.GetBytes($".{password}"))), 0, iv, 0, sizeof(Guid));
            fixed (byte* piv = iv)
            fixed (byte* pkey = key)
                RC4.rc4_crypt(pkey, cipher.KeyLength, piv, ivLen, 0, 0);
        }

        private static byte[] merges(params byte[][] s)
        {
            if (s == null || s.Length <= 0)
                return BufferSegment.Empty;
            byte[] a = BufferSegment.Empty;
            foreach (byte[] i in s)
                a = merge(a, i);
            return a;
        }

        private static byte[] merge(byte[] a, byte[] b)
        {
            int al = a?.Length ?? 0;
            int bl = b?.Length ?? 0;
            int rl = al + bl;
            if (rl <= 0)
                return BufferSegment.Empty;
            byte[] r = new byte[rl];
            Buffer.BlockCopy(a, 0, r, 0, al);
            Buffer.BlockCopy(b, 0, r, al, bl);
            return r;
        }

        public virtual BufferSegment Encrypt(BufferSegment data)
        {
            lock (this.LockObj)
            {
                if (data == null || data.Length <= 0)
                {
                    return new BufferSegment(BufferSegment.Empty);
                }

                int outLen = data.Length + cipher.BlockSize;
                byte[] cipherText = new byte[outLen];
                fixed (byte* buf = &data.Buffer[data.Offset])
                {
                    // INIT-CTX
                    initCipher(ref encryptCTX, iv, true);
                    if (Native.EVP_CipherUpdate(encryptCTX, cipherText, out outLen, buf, data.Length) <= 0)
                    {
                        return new BufferSegment(BufferSegment.Empty);
                    }
                }
                return new BufferSegment(cipherText, outLen);
            }
        }

        public virtual BufferSegment Decrypt(BufferSegment data)
        {
            lock (this.LockObj)
            {
                if (data == null || data.Length <= 0)
                {
                    return new BufferSegment(BufferSegment.Empty);
                }
                
                int outLen = data.Length + cipher.BlockSize;
                byte[] cipherText = new byte[outLen];
                fixed (byte* buf = &data.Buffer[data.Offset])
                {
                    // INIT-CTX
                    initCipher(ref decryptCTX, iv, false);
                    if (Native.EVP_CipherUpdate(decryptCTX, cipherText, out outLen, buf, data.Length) <= 0)
                    {
                        return new BufferSegment(BufferSegment.Empty);
                    }
                }
                return new BufferSegment(cipherText, outLen);
            }
        }
    }
}
